### ISSUE TITLE
Proposal for `raceBoth` function

### ISSUE THREAD

**Issue by Gabriel439** [simonmar/async#38]
I would like to propose adding the following `raceBoth` function if it's alright with the `async` maintainers.  This function runs two `IO` actions at the same time and returns both of their results in the order in which they complete:

```haskell
{-| Run two `IO` actions concurrently, returning their results in the same order
    which the two actions complete
-}
raceBoth :: IO a -> IO b -> IO (Either (a, IO b) (b, IO a))
```

The most common motivating use case for this function is to run some `IO` action concurrently with a timeout, and then doing something if the timeout completes first without canceling the other action.  For example, you might try to read a fixed number of elements from some handle and if the timeout expires you can flush the current number of elements that you've read so far without canceling the current read:

```haskell
example = do
    x <- raceBoth (threadDelay 1000000) readElement
    case x of
        -- Timeout reached!  Flush the elements we collected so far
        -- and then run `io` to finish reading the element
        Left  ((), io) -> ...
        -- Element read in time!  Keep going
        Right (e , _ ) -> ...
```

Here's the code I have been using for this function in my utility modules which is similar to the `race`/`concurrently` code in `async`:

```haskell
import Control.Exception (SomeException, catch, onException)

import qualified Control.Exception       as Exception
import qualified Control.Concurrent      as Concurrent
import qualified Control.Concurrent.MVar as MVar

catchAll :: IO a -> (SomeException -> IO a) -> IO a
catchAll = Exception.catch

{-| Run two `IO` actions concurrently, returning their results in the same order
    which the two actions complete
-}
raceBoth :: IO a -> IO b -> IO (Either (a, IO b) (b, IO a))
raceBoth ioA ioB = do
    mvar <- MVar.newEmptyMVar
    Exception.mask (\restore -> do
        let threadA = do
                a <- ioA
                MVar.putMVar mvar (Right (Left a))
        let threadB = do
                b <- ioB
                MVar.putMVar mvar (Right (Right b))

        let handler e = MVar.putMVar mvar (Left e)
        idA <- Concurrent.forkIO (threadA `catchAll` handler)
        idB <- Concurrent.forkIO (threadB `catchAll` handler)

        let oops = error "raceBoth: the impossible happened!"

        let stopBoth = do
                Concurrent.killThread idB
                Concurrent.killThread idA

        let collectA = do
                e0 <- MVar.takeMVar mvar
                case e0 of
                    Left   exc      -> Exception.throwIO exc
                    Right (Left  a) -> return a
                    _               -> oops
        let collectB = do
                e0 <- MVar.takeMVar mvar
                case e0 of
                    Left   exc      -> Exception.throwIO exc
                    Right (Right b) -> return b
                    _               -> oops
        let collectBoth = do
                e0 <- MVar.takeMVar mvar
                case e0 of
                    Left exc -> do
                        stopBoth
                        Exception.throwIO exc
                    Right e1 -> case e1 of
                        Left  a -> return (Left  (a, collectB))
                        Right b -> return (Right (b, collectA))

        r <- restore collectBoth `onException` stopBoth
        return r )
```

**Comment by simonmar** [comment]
I like the idea, but the semantics is slightly odd because unlike `concurrently` and `race`, `raceBoth` may leak threads.

I like to try to specify these things using the basic combinators, and then implement them in a more efficient way if possible.  `raceBoth` is something like this:

```
raceBoth :: IO a -> IO b -> IO (Either (a, IO b) (b, IO a))
raceBoth a b = do
  aa <- async a
  ab <- async b
  e <- waitEither aa ab
  case e of
    Left ra -> return (Left (ra, wait ab))
    Right rb -> return (Right (rb, wait aa))
```

But that's not quite right, because if an exception is thrown then we kill the threads.  As far as I can tell, there's no equivalent specification of `raceBoth` in terms of the other combinators, which bothers me.  Perhaps we should provide another primitive, something like

```
withAsyncCancelledOnException :: IO a -> (Async a -> IO b) -> IO b
withAsyncCancelledOnException a = bracketOnException (async a) cancel
```

I don't like the naming though.

**Comment by Gabriel439** [comment]
Do we need a new primitive?  Why not make the semantic specification directly in terms of `bracketOnException`?

```haskell
bracketOnException (async ioA) cancel (\asyncA ->
bracketOnException (async ioB) cancel (\asyncB ->
    e <- waitEither asyncA asyncB
    ... ))
```

**Comment by simonmar** [comment]
Yes, we could do that.

I like this version slightly better:

```
withRaceBoth :: IO a -> IO b -> (Either (a, IO b) (b, IO a) -> IO c) -> IO c
withRaceBoth a b inner =
  withAsync a $ \aa ->
  withAsync b $ \ab -> do
  e <- waitEither aa ab
  case e of
    Left ra -> inner (Left (ra, wait ab))
    Right rb -> inner (Right (rb, wait aa))
```

because it doesn't leak threads and doesn't have the unusual `bracketOnException` behaviour.  But would it work for your use cases?

**Comment by Gabriel439** [comment]
Sorry for the very long delay.  So the conclusion I eventually came to was that I did not need this.  Pretty much every use case I could come up with ended up being better served by using a channel and the more I dug into this problem the more I realized that I was essentially reinventing a lot of the machinery that a channel provides.

For example, going back to the original problem that I was trying to solve, I found that it was much easier to just have two threads feeding the channel: one of which would feed in the result I was expecting wrapped in `Just` and a concurrent thread feeding in a `Nothing` if the action timed out before the other thread completed.

So I am actually dropping this request since I don't want to clutter the library with a utility if I can't think of an immediate practical need for it.
